core/lib/zksync_core/src/api_server/web3/pubsub.rs

```diff
@@ -238,118 +238,174 @@ impl EthSubscribe {
         self.events_sender = Some(sender);
     }
 
-    /// Assigns ID for the subscriber if the connection is open, returns error otherwise.
-    fn assign_id(
-        subscriber: typed::Subscriber<PubSubResult>,
-    ) -> Result<(typed::Sink<PubSubResult>, SubscriptionId), ()> {
-        let id = H128::random();
-        let sub_id = SubscriptionId::String(format!("0x{}", hex::encode(id.0)));
-        let sink = subscriber.assign_id(sub_id.clone())?;
-        Ok((sink, sub_id))
+    async fn reject(sink: PendingSubscriptionSink) {
+        sink.reject(ErrorObject::borrowed(
+            ErrorCode::InvalidParams.code(),
+            "Rejecting subscription - invalid parameters provided.",
+            None,
+        ))
+        .await;
+    }
+
+    async fn heads_subscriber(
+        pending_sink: PendingSubscriptionSink,
+        b: broadcast::Receiver<BlockHeader>,
+    ) {
+        let _guard = PUB_SUB_METRICS.active_subscribers[&SubscriptionType::Blocks].inc_guard(1);
+        let sink = pending_sink.accept().await.unwrap();
+        let closed = sink.closed().fuse();
+        let mut b = BroadcastStream::new(b);
+        tokio::pin!(closed);
+
+        loop {
+            tokio::select! {
+                new_log = b.next() => {
+                    if let Some(Ok(new_log)) = new_log {
+                        sink.send_timeout(
+                            SubscriptionMessage::from_json(&PubSubResult::Header(new_log))
+                            .expect("PubSubResult always serializable to json; qed"),
+                            SUBSCRIPTION_SINK_SEND_TIMEOUT,
+                        )
+                            .await.ok();
```

Comment:
- Nit: incorrect formatting (Rust formatter is not very good at formatting inside macros like `tokio::select!` in this case).
- This implies that some events for a subscriber may be dropped, which I don't think is reasonable. Either we should stop the task on a timeout, or use `send()` (i.e., wait indefinitely).

Reply:
formatted